content/blog/elm-at-noredink/index.md

```diff
@@ -0,0 +1,445 @@
+    -- User actions
+    | OpenModal Modal.Kind String
+    | CloseModal
+    | SetModalKind Modal.Kind
+    | ModalError Http.Error
+    | AddStudent Modal.CreateStudentAccountsManuallyModel
+    | Archived Int
+    | Unarchived Int
+    | ToggleTooltip String Bool
+```
+
+## When to nest TEA modules
+
+If you looked at the section above and you thought "Welp, that looks like a
+lot of work", I hear you. As a matter of fact, we don't need to do the
+whole TEA-nesting dance everytime we want to reuse a module. In fact, most of
+the UI components that we use don't have their own state, so they don't
+need even need their own `update` function. Let's look at how we use a
+button:
+
+```elm
+import Nri.Ui.Button.V10 as Button
+
+view : Env -> Model -> Html Msg
+view env model =
+    div []
+        [ Button.link "Edit"
+            [ Button.css [ Css.marginBottom (Css.px 10) ]
+            , Button.onClick EditAssignment
+            , Button.icon UiIcon.calendar
+            , Button.small
+            ]
+        ]
+```
+
+It seems obvious, but if there is no state, then there is no need to do
+extra work. If you are curious on how we structure our UI components, you
+can take a gander at [noredink-ui](https://github.com/NoRedInk/noredink-ui) (or
+check out the preview [here](https://noredink-ui.netlify.app/)). We're
+always including simple examples for each UI component, so that they act as
+the best documentation possible
+([component](https://noredink-ui.netlify.app/#/doodad/Switch) /
+[example](https://github.com/NoRedInk/noredink-ui/blob/master/styleguide-app/Examples/Switch.elm)).
+
+In a way, if you feel that it is too complex to nest TEA applications,
+it's because it is. I would recommend reaching for it as a last resort,
+like using a big bazooka to plant daffodils in your garden: it works, but
+it's a bit heavy handed.
+
+Instead you might find some abstractions which don't require another model
+and another set of update messages. You can do lots with a view function
+that takes a config record and an extensible record:
+
+```elm
+viewSidebar :
+    { onClick : msg, onClose : msg }
+    -> { a | sidebarItems : List SidebarItem }
+    -> Html msg
+```
+
+As mentioned above, choose the simplest solution that works. Later you can
+refactor and make it feel good.
+
+## Interacting with JavaScript
+
+We mainly use [ports](https://guide.elm-lang.org/interop/ports.html) to
+interact with JavaScript components. We have two simple rules for ports:
+- All ports return values as `Json.Value`. This is one of the simplest ways
+  to introduce runtime errors in your application. By treating all values
+  as JSON blobs, we must decode them and deal with the eventual decoding
+  failure.
+- All port functions must be documented in the Elm module. We don't want to
+  go hunting in JavaScriptLand how a port is being used. A couple of
+  lines explaining what the port triggers can go a long way.
+
+In cases where ports are not enough, we use [custom
```

Comment:
I know there's a lot to be said about this and you can't fit it all in a sentence, but to me "when ports are not enough" gives the idea that custom elements are a strictly more powerful way to interact with JS.

Do you think there's a succinct and more accurate way to to describe what makes a good use case for custom elements over ports?

([Luke's talk](https://www.youtube.com/watch?v=tyFe9Pw6TVE&t=12s&ab_channel=ElmEurope) might be a good reference here, but maybe you don't want custom elements to take up so much space).

Reply:
Cool, I'll try to clarify this :)